### DISCUSSION TITLE
Transform Precoding (DFT-s-OFDM) in PUSCH

### DISCUSSION THREAD

**Discussion by danielschaeufele** [NVlabs/sionna#268] [General]
I'm currently doing some experiments with DFT-spread-OFDM in the PUSCH. There is a `PUSCHConfig.transform_precoding` parameter, which should control this, but it is never used, so my conclusion is, that this isn't implemented.

I'm interested in contributing and implementing this myself, but I'm not 100% sure how this feature should be best implemented. As far as I can see only the RessourceGridMapper has access to the information on which data symbol will be mapped onto which OFDM symbol. However, logically it seems to be a bit weird to me. Do you know a better place to implements this or should I add it to the RessourceGridMapper?

**Comment by jhoydis** [comment]
Hi, 

DFT-spread-OFDM is indeed not implemented yet (and I have not looked into the standard for details). My understanding would be that this boils down to a small DFT on a few adjacent symbols in the frequency domain. If this is the case, this would happen prior to the mapping onto the resource grid. Do you agree?

**Reply by danielschaeufele** [reply]
Yes exactly. The problem is, that the DFT is performed per OFDM symbol. So all data symbols need to be separated by OFDM symbol, then the DFT can be computed and then the mapping onto the resource grid can be done. If I'm not missing something then a separate class would need to duplicate some functionality from the ResourceGridMapper.

**Reply by jhoydis** [reply]
The data symbols are mapped to the resource grid in the order top-bottom-left-right (i.e., OFDM symbol by OFDM symbol). If one would reshape them to a matrix of dimensions [DFT-size, -1] one could apply the DFT from the left and then reshape the matrix back to a vector. The only thing which needs to be known for this is the size of the DFT. The resulting vector could then be fed into the ResourceGridMapper, no?

**Reply by danielschaeufele** [reply]
Yes, but this assumes that every OFDM symbol has the exact same number of data symbols. Is this always the case?

**Reply by jhoydis** [reply]
Actually, I am not even sure if the precoding is done before or after mapping to the resource grid. Need to check the standard before continuing the discussion. The number of data symbols is the same for the OFDM symbols that do not carry DMRS.

**Reply by jhoydis** [reply]
After a quick look at TS 38.211 and TS 38.214, it seems that when transform precoding is enabled, 
- there is only a single layer,
- DMRS and data are not multiplexed, i.e., numCDMGroupsWithoutData=2.

This means that the approach described above should work. You simply reshape the vector with data symbols to a matrix of size [number of allocated subcarriers, -1] and then apply the DFT from the left. That means that one could think about a PUSCHTransformPrecoder class that would get a list of punch_configs and that would sit between the LayerMapper and ResourceGridMapper.

Apart from this one would need to look at DMRS generation when transform precoding is enabled (it seems that low-PAPR sequences are used) and consider implementing PI/2-BPSK support for Mapper and Demapper.

**Reply by danielschaeufele** [reply]
Thank you for the advise. I will take a closer look at the standard and then implement it like you proposed.

**Reply by danielschaeufele** [reply]
The implementation for the transmitter seems to be pretty straight-forward so far. However, implementing the receiver seems to be much harder.

If I understand DFT-s-OFDM correctly, decoding works in the following way:
1. FFT computation to go from time domain to frequency domain
2. equalization in the frequency domain
3. IDFT computation
4. demapping and all the other steps from normal CP-OFDM

I think it is impossible to build a generic implementation that works for all detectors implemented by Sionna, so my proposal is to implement this only for the linear detector. To do that, I would either add a parameter to `LinearDetector` or create a new class `LinearTransformPrecodingDetector`. The first option has the disadvantage, that the generic MIMO detector has a dependency on a 5G NR PUSCH function, which seems to be really bad encapsulation. The second option has a bit of code duplication, as only a few lines would need to be added to `LinearDetector`. I'm unsure, which way is best or if there are even better methods, that I didn't think of.

**Reply by jhoydis** [reply]
Hm, if I understand it correctly, the only difference between a detector for OFDM and DFT-s-OFDM is that we have the FFT step prior to demapping for the latter. I think that there is no need to have an implementation for all detectors because DFT-s-OFDM is mainly used for single-stream transmissions.

I like your suggestion to create a ``LinearTransformPrecodingDetector`` class. However, I think that all linear detectors would perform equally for a single stream. The question is if there is any value in supporting MIMO detection for DFT-s-OFDM?

The PUSCH receiver would then need to select this new ``TransformPrecodingDetector`` if transform precoding is selected.

**Comment by danielschaeufele** [comment]
I finally managed to find time to continue with the implementation, sorry for the long delay.

I pushed my implementation here in case you already want to have a look: https://github.com/danielschaeufele/sionna/tree/transform_precoding

Currently I'm still missing documentation and tests. After I added those I will prepare a pull request.

I opted against implementing Pi/2-BPSK, because this is optional in the standard and it would need significant modification to the Mapper and Demapper.

**Reply by danielschaeufele** [reply]
Regarding the tests for the `PUSCHTransmitter`: I would like to add a new test case. Are there any instructions on how to do that and on how the test data is generated?

**Reply by SebastianCa** [reply]
There are no detailed instructions available for generating test cases, but they should cover all relevant scenarios (ideally some corner cases). Please ensure that the test cases are not too computationally demanding (e.g., they should run in a few seconds) to avoid excessive runtime for all unit tests.

For data generation, you can either test against a small dataset of pre-calculated cases or use a NumPy reference implementation.

**Reply by danielschaeufele** [reply]
For tests that only test a single module, I will do that (e.g. the tests for the Zadoff-Chu sequences are tested against the matlab implementation). However, I think it also makes sense to test the whole `PUSCHTransmitter` chain. As far as I can see, the tests in `test/unit/nr/test_pusch_transmitter.py` use predefined test files. The JSON files seem to be generated by `test/unit/nr/pusch_test_configs/gen_test_configs.py`, but there is no reference as to how the `npy` files have been generated. Additionally, adding a new loop for `transform_precoding` turned on/off is not going to work, because the parameters that can be used when `transform_precoding` is enabled are severely restricted (e.g. only a single layer is supported). So maybe it would make sense to add just very specific tests manually. Is this supported or will it break automatic generation in the future?

**Reply by SebastianCa** [reply]
It should be fine to add an additional "manual" test case for the supported set of parameters. However, please ensure that `make test` runs these cases automatically (which should be the case according to how I understand your question).

Btw: please ensure that non-supported configurations are detected during the initialization.

**Reply by danielschaeufele** [reply]
I'm finished with my implementation, but I'm having some issues with running the tests. First of all there seems to be a memory leak, as 32GiB of RAM are full after running ~20 tests and the tests will abort. My current workaround is to run all tests individually.

The second issue is that the test in `integration/test_ofdm_mimo_estimation_detection.py` fails even when I'm in the main branch. Does this happen only on my system? I'm running on a server without a GPU, so maybe this is an issue? This is the test output:

```
-> % pytest integration/test_ofdm_mimo_estimation_detection.py
================================================================================================ test session starts ================================================================================================
platform linux -- Python 3.10.12, pytest-8.2.0, pluggy-1.5.0
rootdir: /home/daniel/python-test/sionna
configfile: pyproject.toml
plugins: anyio-4.2.0
collected 1 item

integration/test_ofdm_mimo_estimation_detection.py F                                                                                                                                                          [100%]

===================================================================================================== FAILURES ======================================================================================================
_______________________________________________________________________________ TestOFDMMIMODetectors.test_all_detectors_in_all_modes _______________________________________________________________________________

self = <test_ofdm_mimo_estimation_detection.TestOFDMMIMODetectors testMethod=test_all_detectors_in_all_modes>

    def test_all_detectors_in_all_modes(self):
        """Test for all detectors in all execution modes
        """

        tf.random.set_seed(42)

        for detector in ["lmmse", "ep", "k-best", "mmse-pic"]:
            for output in ["bit", "symbol"]:
                model = MIMOOFDMLink(output, detector, False, 4, 2)
                # Eager
>               er_eager = compute_ber(*model(1, 40.0))

integration/test_ofdm_mimo_estimation_detection.py:213:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../.venv/lib/python3.10/site-packages/keras/src/utils/traceback_utils.py:70: in error_handler
    raise e.with_traceback(filtered_tb) from None
integration/test_ofdm_mimo_estimation_detection.py:188: in call
    det_out = self.detector((y_rg,h_hat,prior,err_var,no))
../sionna/ofdm/detection.py:522: in call
    z = self._detector([y_dt, h_dt_desired, prior, s])
../sionna/mimo/detection.py:1946: in call
    llr_d, llr_a, _ = tf.while_loop(dec_stop,
../sionna/mimo/detection.py:1863: in mmse_pic_self_iteration
    x_hat, var_x = self._symbol_logits_2_moments(x_logits)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <sionna.mapping.SymbolLogits2Moments object at 0x7f8a0c72e800>
logits = <tf.Tensor: shape=(1, 1, 14, 48, 4, 4), dtype=float32, numpy=
array([[[[[[-1.3862944, -1.3862944, -1.3862944, -1.38629...62944, -1.3862944, -1.3862944],
           [-1.3862944, -1.3862944, -1.3862944, -1.3862944]]]]]],
      dtype=float32)>

    def __call__(self, logits):
>       p = tf.math.softmax(logits, axis=-1)
E       tensorflow.python.framework.errors_impl.AbortedError: Exception encountered when calling layer 'mmsepic_detector' (type MMSEPICDetector).
E
E       {{function_node __wrapped__Softmax_device_/job:localhost/replica:0/task:0/device:CPU:0}} Input dims must be <= 5 and >=1 [Op:Softmax] name:
E
E       Call arguments received by layer 'mmsepic_detector' (type MMSEPICDetector):
E         • inputs=['tf.Tensor(shape=(1, 1, 14, 48, 16), dtype=complex64)', 'tf.Tensor(shape=(1, 1, 14, 48, 16, 4), dtype=complex64)', 'tf.Tensor(shape=(1, 1, 14, 48, 4, 2), dtype=float32)', 'tf.Tensor(shape=(1, 1
, 14, 48, 16, 16), dtype=complex64)']

../sionna/mapping.py:1531: AbortedError
----------------------------------------------------------------------------------------------- Captured stderr call ------------------------------------------------------------------------------------------------
2024-06-04 13:26:35.439256: I external/local_xla/xla/service/service.cc:168] XLA service 0x557038f2e1d0 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2024-06-04 13:26:35.439322: I external/local_xla/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
I0000 00:00:1717507595.459299   28896 device_compiler.h:186] Compiled cluster using XLA!  This line is logged at most once for the lifetime of the process.
2024-06-04 13:26:35.459560: E external/local_xla/xla/stream_executor/stream_executor_internal.h:177] SetPriority unimplemented for this stream.
2024-06-04 13:26:40.795246: E external/local_xla/xla/stream_executor/stream_executor_internal.h:177] SetPriority unimplemented for this stream.
2024-06-04 13:26:49.062902: W tensorflow/core/framework/op_kernel.cc:1839] OP_REQUIRES failed at mkl_softmax_op.cc:252 : ABORTED: Input dims must be <= 5 and >=1
------------------------------------------------------------------------------------------------- Captured log call -------------------------------------------------------------------------------------------------
WARNING  tensorflow:math_ops.py:1017 You are casting an input of type complex64 to an incompatible dtype float32.  This will discard the imaginary part and may not be what you intended.
WARNING  tensorflow:math_ops.py:1017 You are casting an input of type complex64 to an incompatible dtype float32.  This will discard the imaginary part and may not be what you intended.
================================================================================================= warnings summary ==================================================================================================
../../.venv/lib/python3.10/site-packages/tensorflow/__init__.py:30
  /home/daniel/python-test/.venv/lib/python3.10/site-packages/tensorflow/__init__.py:30: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 6
32 for potential alternatives
    import distutils as _distutils

../../.venv/lib/python3.10/site-packages/traitlets/__init__.py:26
  /home/daniel/python-test/.venv/lib/python3.10/site-packages/traitlets/__init__.py:26: DeprecationWarning:
              Sentinel is not a public part of the traitlets API.
              It was published by mistake, and may be removed in the future.

    warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============================================================================================== short test summary info ==============================================================================================
FAILED integration/test_ofdm_mimo_estimation_detection.py::TestOFDMMIMODetectors::test_all_detectors_in_all_modes - tensorflow.python.framework.errors_impl.AbortedError: Exception encountered when calling layer 'm
msepic_detector' (type MMSEPICDetector).
========================================================================================== 1 failed, 2 warnings in 49.03s ===========================================================================================
```

**Reply by danielschaeufele** [reply]
Good news, I just tried on our compute cluster while running everything in the docker container (or more precisely the equivalent apptainer container) and the failed tests are gone. So it seems to be an issue with either running on CPU or some of the library versions. I will squash all commits now and then create a pull request.

**Reply by SebastianCa** [reply]
Hi @danielschaeufele,

indeed, running all tests together requires quite some GPU memory. 

Thank you again for the PR. We continue the discussion about this PR in #459.
